File: gtci/06-in-place-reversal-of-linked-list/cpp/0-reverse-a-linked-list.cpp
#include <cassert>
#include <iostream>

#include "linked_list_utils.h"

using namespace std;

class Solution {
   public:
    ListNode* reverse(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* curr = head;

        while (curr != nullptr) {
            ListNode* temp = curr->next;

            curr->next = prev;
            prev = curr;

            curr = temp;
        }

        return prev;
    }
};

int main() {
    Solution sol;
    bool allTestsPassed = true;

    // Test Case 1: Multiple nodes (1->2->3 should become 3->2->1)
    {
        ListNode* head = new ListNode(1);
        head->next = new ListNode(2);
        head->next->next = new ListNode(3);

        ListNode* reversed = sol.reverse(head);
        vector<int> result = getLinkedListValues(reversed);
        vector<int> expected = {3, 2, 1};

        assert(result == expected &&
               "Test Case 1 Failed: Multiple nodes not reversed correctly");
        cleanupLinkedList(reversed);
    }

    // Test Case 2: Single node (should remain the same)
    {
        ListNode* singleNode = new ListNode(5);
        ListNode* reversed = sol.reverse(singleNode);
        vector<int> result = getLinkedListValues(reversed);
        vector<int> expected = {5};

        assert(result == expected &&
               "Test Case 2 Failed: Single node case failed");
        cleanupLinkedList(reversed);
    }

    // Test Case 3: Empty list (nullptr should remain nullptr)
    {
        ListNode* emptyList = nullptr;
        ListNode* reversed = sol.reverse(emptyList);
        assert(reversed == nullptr &&
               "Test Case 3 Failed: Empty list should return nullptr");
    }

    cout << "All test cases passed." << endl;
    return 0;
}

File: gtci/06-in-place-reversal-of-linked-list/python3/0-reverse-a-linked-list.py
class Node:
    def __init__(self, value, next=None):
        self.val = value
        self.next = next


class Solution:
    def reverse(self, head):
        prev = None
        curr = head

        while curr is not None:
            temp = curr.next

            curr.next = prev
            prev = curr

            curr = temp

        return prev


if __name__ == "__main__":
    from util import get_linked_list_values

    # Test case 1: Multiple nodes (1->2->3 should become 3->2->1)
    head = Node(1)
    head.next = Node(2)
    head.next.next = Node(3)

    sol = Solution()
    reversed_head = sol.reverse(head)
    result = get_linked_list_values(reversed_head)
    assert result == [3, 2, 1], f"Expected [3, 2, 1], but got {result}"

    # Test case 2: Single node (should remain the same)
    single_node = Node(5)
    reversed_single = sol.reverse(single_node)
    result = get_linked_list_values(reversed_single)
    assert result == [5], f"Expected [5], but got {result}"

    # Test case 3: Empty list (None should remain None)
    empty_list = None
    reversed_empty = sol.reverse(empty_list)
    assert reversed_empty is None, "Expected None for empty list"

    print("All test cases passed.")
